### DISCUSSION TITLE
Question about the zerolatency option

### DISCUSSION THREAD

**Discussion by beosliege** [nagadomi/nunif#102] [Q&A] [answered]
Hello, 

My experience with "zero latency" tells me that it is a feature used in live streaming to reduce any lag time in the video encoding. Is this correct, and if so, how does that apply here for 3d video conversion? I would not imagine that latency would be an issue in this context?

Thank you for this project. I am really enjoying it a lot!

**Comment by nagadomi** [comment]
I am not very familiar with video encoding.

Encoded videos are usually streamed from network drives. Also, when outputting mkv file, they are streamed live during encoding.
I had set it to default thinking it would have something to do with that. 
Also, I do not feel that anything has changed as a result of that setting.

Your opinion is that it should be turned off by default?
Regarding having option itself,  I just provide all available options.

**Reply by beosliege** [reply]
When I search the web for the FFMPEG's zerolatency option, I primarily receive discussions about live-streaming: For example, webcams, gaming, or other live broadcasts where any delay in the video encoding would not be favorable to the user that is watching the video while it is being encoded in real-time. This is what lead me to believe the zerolatency option is for live streaming, instead of pre-encoded videos like what we are doing here in iw3.

There are other instances where minimizing lag time during encoding is useful, such as for surveillance systems or for video editing applications. In a sense, these are also "live streaming" where you do not want any delay in the real-time encoding-to-delivery.

In one discussion (https://stackoverflow.com/questions/16658873/how-to-minimize-the-delay-in-a-live-streaming-with-ffmpeg), "-tune zerolatency" is mentioned because it helped reduce latency during a live stream, but it also "made the qualiity of the motion vectors a bit worst [sic]".

The ffmpeg documentation (https://trac.ffmpeg.org/wiki/Encode/H.264) mentioned low latency in two areas:

"zerolatency – good for fast encoding and low-latency streaming"

"x264 offers a -tune zerolatency option for low latency streaming."

I wish the FFMPEG documentation were a bit more clear on the details. However, it seems from general consensus that zerolatency is primarily aimed at users who are encoding and delivering video in real-time. Whereas, what we are doing here with iw3 is to spend a long time to create a very high-quality encode of a video that is to be played back at a later time once it is finished (not streamed in real-time). For this reason, I would suggest disabling zerolatency as a default option.

I also noticed that you offer the "fast decode" option. I believe this could potentially be useful for users where the final video will be running on lower-end hardware, or perhaps a system where hardware acceleration is not available: This option changes settings to prioritize faster / easier decoding of the final encoded video, but potentially at the expense of video quality. It is a trade-off, but it is an option that could be useful to some people.

I am not an expert on FFMPEG, but if you find my research reasonable, I hope it will help. I am open to any correction as I learn more.

Thanks!

**Reply by nagadomi** [reply, marked as answer]
Thank you for the info.
I did a quick search after this discussion was posted and I agree with you.
Also, as noted above, I don't see anything changing whether this option is off or on, so I have no objection to changing the default value.

I will change zerolatency option to off by default.

**Comment by nagadomi** [comment]
Changed.
GUI input state is stored in `nunif/tmp/iw3-gui.cfg`, you can check the default values by deleting this file.
